app.py

```diff
@@ -32,21 +32,79 @@
     'upstream/Rover-4.2' : 'Rover 4.2 stable'
 }
 default_branch = 'upstream/master'
-CURR_BRANCH = default_branch
 
-def get_boards_from_ardupilot_tree(branch):
+def get_vehicles():
+    return VEHICLES
+
+def get_default_vehicle():
+    return default_vehicle
+
+def get_branches():
+    return BRANCHES
+
+def get_default_branch():
+    return default_branch
+
+# LOCKS
+queue_lock = Lock()
+head_lock = Lock()  # lock git HEAD, i.e., no branch change until this lock is released
+
+def is_valid_branch(branch):
+    return get_branches().get(branch) is not None
+
+def get_git_hash(branch):
+    return subprocess.check_output(['git', 'rev-parse', branch], cwd=sourcedir, encoding='utf-8', shell=False).rstrip()
+
+def on_branch(branch):
+    git_hash_target = get_git_hash(branch)
+    app.logger.info("Expected branch git-hash '%s'" % git_hash_target)
+    git_hash_current = get_git_hash('HEAD')
+    app.logger.info("Current branch git-hash '%s'" % git_hash_current)
+    return git_hash_target == git_hash_current
+
+def checkout_branch(targetBranch, s_dir, fetch_and_reset=False, temp_branch_name=None):
+    '''checkout to given branch and return the git hash'''
+    # Note: remember to acquire head_lock before calling this method
+    if not is_valid_branch(targetBranch):
+        app.logger.error("Checkout requested for an invalid branch")
+        return None 
+    remote =  targetBranch.split('/', 1)[0]
+    branch_name = targetBranch.split('/', 1)[1]
+    if not on_branch(targetBranch):
+        app.logger.info("Checking out to %s branch" % targetBranch)
+        subprocess.run(['git', 'checkout', targetBranch], cwd=s_dir, shell=False)
```

Comment:
This line is responsible for checkout. And here `targetBranch` is already in format <remote>/<branch_name>

Was this what you were looking for @MichelleRos, @peterbarker ? Please correct me if I am getting something wrong.

Reply:
Ah, right.  I didn't read through all the logic properly.  For some reason we're doing multiple checkouts, depending on whether you're sending through a target branch name.  I thought this flow was always deleting branches when I glanced through it.